src/x509.rs

```diff
@@ -4,28 +4,46 @@
 //!
 
 use std::fmt;
+use std::collections::HashMap;
 
 use num_bigint::BigUint;
 use time::Tm;
 
-use der_parser::ber::BitStringObject;
-use der_parser::der::DerObject;
-use der_parser::oid::Oid;
-use objects::{oid2nid,nid2sn};
-use error::X509Error;
-use x509_parser::parse_ext_basicconstraints;
+use der_parser::{
+    ber::BitStringObject,
+    der::DerObject,
+    oid,
+    oid::Oid,
+};
+use crate::objects::{oid2nid,nid2sn};
+use crate::error::X509Error;
+use crate::x509_extensions;
 
 
 #[derive(Debug, PartialEq)]
 pub struct X509Extension<'a> {
-    pub oid:  Oid,
+    pub oid:  Oid<'a>,
     pub critical: bool,
     pub value: &'a[u8],
+    pub(crate) extension_type: Option<x509_extensions::ExtensionType<'a>>,
```

Comment:
Not sure I get the intention of this. Is it a way to optionally parse and store parsed extensions (since parsing can be expensive) ? It could be nice to deep-parse only if requested, and store result.